providers/Spotify/mod.ts

```diff
@@ -0,0 +1,318 @@
+				'Content-Type': 'application/x-www-form-urlencoded',
+			},
+			body: body,
+		});
+
+		const content = await response.json();
+		return {
+			accessToken: content?.access_token,
+			validUntilTimestamp: Date.now() + (content.expires_in * 1000),
+		};
+	}
+}
+
+export class SpotifyReleaseLookup extends ReleaseApiLookup<SpotifyProvider, Album> {
+	constructReleaseApiUrl(): URL {
+		let lookupUrl: URL;
+		const query = new URLSearchParams();
+		if (this.lookup.method === 'gtin') {
+			lookupUrl = new URL(`search`, this.provider.apiBaseUrl);
+			query.set('type', 'album');
+			query.set('q', `upc:${this.lookup.value}`);
+		} else { // if (method === 'id')
+			lookupUrl = new URL(`albums/${this.lookup.value}`, this.provider.apiBaseUrl);
+		}
+
+		lookupUrl.search = query.toString();
+		return lookupUrl;
+	}
+
+	protected async getRawRelease(): Promise<Album> {
+		if (this.lookup.method === 'gtin') {
+			// Spotify does not always find UPC barcodes but expects them prefixed with
+			// 0 to a max. size of 14 characters. E.g. "810121774182" gives no results,
+			// but "00810121774182" does.
+			let albumId: string | undefined;
+			while (this.lookup.value.length <= 14 && !albumId) {
+				const cacheEntry = await this.provider.query<SearchResult>(
+					this.constructReleaseApiUrl(),
+					this.options.snapshotMaxTimestamp,
+				);
+				if (cacheEntry.content?.albums?.items?.length) {
+					albumId = cacheEntry.content.albums.items[0].id;
+				}
+				// Prefix the GTIN with an additional 0
+				this.lookup.value = `0${this.lookup.value}`;
+			}
+
+			// No results found
+			if (!albumId) {
+				throw new ResponseError(this.provider.name, 'API returned no results', this.constructReleaseApiUrl());
+			}
+
+			// Result is a SimplifiedAlbum. Perform a regular ID lookup with the found release
+			// ID to retrieve complete data.
+			this.lookup.method = 'id';
+			this.lookup.value = albumId;
+		}
+
+		const cacheEntry = await this.provider.query<Album>(
+			this.constructReleaseApiUrl(),
+			this.options.snapshotMaxTimestamp,
+		);
+		const release = cacheEntry.content;
+
+		this.updateCacheTime(cacheEntry.timestamp);
+		return release;
+	}
+
+	private async getRawTracklist(rawRelease: Album): Promise<Track[]> {
+		let allTracks: SimplifiedTrack[] = [];
+		allTracks = allTracks.concat(rawRelease.tracks.items);
+
+		// The initial response contains max. 50 tracks. Fetch the remaining
+		// tracks with separate requests if needed.
+		let nextUrl = rawRelease.tracks.next;
+		while (nextUrl && allTracks.length < rawRelease.tracks.total) {
+			const cacheEntry = await this.provider.query<ResultList<SimplifiedTrack>>(
+				new URL(nextUrl),
+				this.options.snapshotMaxTimestamp,
+			);
+			this.updateCacheTime(cacheEntry.timestamp);
+			allTracks = allTracks.concat(cacheEntry.content.items);
+			nextUrl = cacheEntry.content.next;
+		}
+
+		// Load full details including ISRCs
+		return this.getRawTrackDetails(allTracks);
```

Comment:
The extra requests should only happen if `this.options.withISRC` is true.
Currently only Deezer implements these "with" flags properly, but I plan to use these to optimize requests in the future, i.e. only enable them if a property is missing and there is no provider which can get them "cheaper".

This would lead to ISRCs not being displayed for Spotify on the release lookup page, but we can change the default options for now, it should not cause any other provider to do additional requests it does not do already (Deezer).
I even just realized that the ISRCs from Spotify would not even be fetched once the release actions page is loaded since the flag is missing from the options there, whoops.

Reply:
Done in 7d0f1dddd96c38e33004742553aca87ab9329a6c